### PR TITLE
LPS-132718 Change apple push notification library from java-apns to pushy

### PR DESCRIPTION
**Pull request by LuismiBarcos**
Liferay used the library [java-apn](https://github.com/notnoop/java-apns) to send apple push notifications. But, as we can see in the README of the repository, the library doesn't work since March 31, 2021. In the same README, it recommends using the library [Pushy](https://github.com/jchambers/pushy) as the new way to send push notifications.

With this PR we do the migration from one library to the other. The main steps are:
* Remove java-apns dependency and add the new library excluding all the packages that are no necessary.
* Delete the code related to the previous library and implement a new way to build the payload and handle the push notifications response.
* Add a new input in the apple push notification configuration to request the AppId

**Main concerns**
I have doubts about how I manage the notifications responses. Basically, it sends always a response, and depending on notification was accepted by the APNs services the response will contain throw an error or not.

Also, a specific doubt about the visibility of the method `sendResponse`. I see that the previous implementation was protected but now, for me that not makes sense, I think it could be a private method. I put it as `protected` just in case there is a reason and I don't see it.

### CODE REVIEW

**Comment by LuismiBarcos** [comment]
sf? Seriously? 🤦 🤦 🤦 
I check it right now

**Comment by LuismiBarcos** [comment]
ci:test:sf

**Comment by LuismiBarcos** [comment]
ci:test:relevant

File: modules/apps/push-notifications/push-notifications-sender/push-notifications-sender-apple/src/main/java/com/liferay/push/notifications/sender/apple/internal/ApplePushNotificationsSender.java [hunk 1]
```diff
@@ -62,23 +72,35 @@
...
 				"Apple push notifications sender is not configured properly");
 		}
 
 		String payload = buildPayload(payloadJSONObject);
 
-		_apnsService.push(tokens, payload);
+		Stream<String> tokensStream = tokens.stream();
+
+		tokensStream.map(
+			t -> new SimpleApnsPushNotification(
```
Comment (nhpatt) [inline modules/apps/push-notifications/push-notifications-sender/push-notifications-sender-apple/src/main/java/com/liferay/push/notifications/sender/apple/internal/ApplePushNotificationsSender.java]: this should be token I think

File: modules/apps/push-notifications/push-notifications-sender/push-notifications-sender-apple/src/main/java/com/liferay/push/notifications/sender/apple/internal/ApplePushNotificationsSender.java [hunk 2]
```diff
@@ -62,23 +72,35 @@
...
+			notification -> _handleNotificationResponse(
+				_apnsClient.sendNotification(notification))
+		);
 	}
 
 	@Activate
 	@Modified
-	protected void activate(Map<String, Object> properties) {
+	protected void activate(Map<String, Object> properties)
+		throws SSLException {
```
Comment (nhpatt) [inline modules/apps/push-notifications/push-notifications-sender/push-notifications-sender-apple/src/main/java/com/liferay/push/notifications/sender/apple/internal/ApplePushNotificationsSender.java:99]: We usually throw Exception

File: modules/apps/push-notifications/push-notifications-sender/push-notifications-sender-apple/src/main/java/com/liferay/push/notifications/sender/apple/internal/ApplePushNotificationsSender.java [hunk 3]
```diff
@@ -108,106 +130,115 @@ protected void activate(Map<String, Object> properties) {
...
 
-		_apnsService = appleServiceBuilder.build();
+		_apnsClient = apnsClientBuilder.build();
 	}
 
 	protected String buildPayload(JSONObject payloadJSONObject) {
-		PayloadBuilder builder = PayloadBuilder.newPayload();
+		SimpleApnsPayloadBuilder builder = new SimpleApnsPayloadBuilder();
 
 		String title = payloadJSONObject.getString(
```
Comment (nhpatt) [inline modules/apps/push-notifications/push-notifications-sender/push-notifications-sender-apple/src/main/java/com/liferay/push/notifications/sender/apple/internal/ApplePushNotificationsSender.java:157]: It's possible that you have to sort the builder

File: modules/apps/push-notifications/push-notifications-sender/push-notifications-sender-apple/src/main/java/com/liferay/push/notifications/sender/apple/internal/messaging/ApplePushNotificationsResponseMessageListener.java [hunk 4]
```diff
@@ -46,33 +34,7 @@ protected void doReceive(Message message) throws Exception {
 			return;
 		}
 
-		AppleResponse appleResponse = (AppleResponse)response;
```
Comment (nhpatt) [inline modules/apps/push-notifications/push-notifications-sender/push-notifications-sender-apple/src/main/java/com/liferay/push/notifications/sender/apple/internal/messaging/ApplePushNotificationsResponseMessageListener.java:49]: Why did you remove this?

Comment (LuismiBarcos) [reply modules/apps/push-notifications/push-notifications-sender/push-notifications-sender-apple/src/main/java/com/liferay/push/notifications/sender/apple/internal/messaging/ApplePushNotificationsResponseMessageListener.java:49]: `DeliveryError.INVALID_TOKEN` was of the java-apns library. I looked for a way to handle Exceptions with Pushy but AFAIK it only throws exceptions, it doesn't handle it.
So, I decided to throw a `PushNotificationException` with the status of the response.

File: modules/apps/push-notifications/push-notifications-sender/push-notifications-sender-apple/src/main/java/com/liferay/push/notifications/sender/apple/internal/ApplePushNotificationsSender.java (hunk 1 above)
Comment (LuismiBarcos) [reply modules/apps/push-notifications/push-notifications-sender/push-notifications-sender-apple/src/main/java/com/liferay/push/notifications/sender/apple/internal/ApplePushNotificationsSender.java]: As you wish 😁

File: modules/apps/push-notifications/push-notifications-sender/push-notifications-sender-apple/src/main/java/com/liferay/push/notifications/sender/apple/internal/messaging/ApplePushNotificationsResponseMessageListener.java (hunk 4 above)
Comment (nhpatt) [reply modules/apps/push-notifications/push-notifications-sender/push-notifications-sender-apple/src/main/java/com/liferay/push/notifications/sender/apple/internal/messaging/ApplePushNotificationsResponseMessageListener.java:49]: 👍

File: modules/apps/push-notifications/push-notifications-sender/push-notifications-sender-apple/src/main/java/com/liferay/push/notifications/sender/apple/internal/ApplePushNotificationsSender.java (hunk 3 above)
Comment (LuismiBarcos) [reply modules/apps/push-notifications/push-notifications-sender/push-notifications-sender-apple/src/main/java/com/liferay/push/notifications/sender/apple/internal/ApplePushNotificationsSender.java:157]: What do you mean by "sort the builder"? Set the attributes alphabetically?

Comment (nhpatt) [reply modules/apps/push-notifications/push-notifications-sender/push-notifications-sender-apple/src/main/java/com/liferay/push/notifications/sender/apple/internal/ApplePushNotificationsSender.java:157]: Yes, the attributes that you are setting like builder.setTitle should be called alfabetically.

Comment (LuismiBarcos) [reply modules/apps/push-notifications/push-notifications-sender/push-notifications-sender-apple/src/main/java/com/liferay/push/notifications/sender/apple/internal/ApplePushNotificationsSender.java:187]: Done!

File: modules/apps/push-notifications/push-notifications-sender/push-notifications-sender-apple/src/main/java/com/liferay/push/notifications/sender/apple/internal/ApplePushNotificationsSender.java (hunk 2 above)
Comment (LuismiBarcos) [reply modules/apps/push-notifications/push-notifications-sender/push-notifications-sender-apple/src/main/java/com/liferay/push/notifications/sender/apple/internal/ApplePushNotificationsSender.java]: Done! (My bad 😞 )

File: modules/apps/push-notifications/push-notifications-sender/push-notifications-sender-apple/src/main/java/com/liferay/push/notifications/sender/apple/internal/ApplePushNotificationsSender.java (hunk 1 above)
Comment (LuismiBarcos) [reply modules/apps/push-notifications/push-notifications-sender/push-notifications-sender-apple/src/main/java/com/liferay/push/notifications/sender/apple/internal/ApplePushNotificationsSender.java]: Done!

**Comment by LuismiBarcos** [comment]
FYI @nhpatt @javierdearcos I've added two commits:
* One to ensure that the `apnsClient` is fully closed. When I only called to close method, it returned a `CompletableFuture` that would be marked as completed when all connections were close and all persistent resources released. So now I wait until this happens to avoid exceptions in the deactivate method (as the CI complain previously)
* Second one to sort the setters of the attributes of the payload builder

**Comment by LuismiBarcos** [comment]
ci:test:sf

**Comment by LuismiBarcos** [comment]
ci:test:relevant

**Comment by nhpatt** [comment]
ci:forward

**Review by nhpatt — APPROVED** [review]
Great job 👏

**Comment by LuismiBarcos** [comment]
I do not know what is happening. In local I am not able to reproduce the error, the deactivate method does not throw any exception. Maybe I have to enable more logs levels in my Liferay to see if there is something I am missing (my developer mode is enabled, of course) 🤔

**Comment by nhpatt** [comment]
Parece que _apnsClient es nulo, no? Y solo se inicializa si tienes un certificado válido, no? Así que están fallando los casos en los que no llegó a inicializarse pero tu en el deactivate asumes que siempre se consigue.

if (_apnsClient != null) ...

**Comment by LuismiBarcos** [comment]
Ay dios mioo!!! 🤦 🤦 🤦

**Comment by LuismiBarcos** [comment]
ci:forward

**Review by javierdearcos — APPROVED** [review]

**Comment by LuismiBarcos** [comment]
Rebase and forward

**Comment by LuismiBarcos** [comment]
ci:forward
